### DISCUSSION TITLE
How to configure watch directory from `config.yaml`, CLI or Bash script?

### DISCUSSION THREAD

**Discussion by a-t-0** [Tribler/tribler#6934] [Q&A] [answered]
Thank you for your marvellous work on Tribler! While trying to configure Tribler with a Bash script, I experienced some difficulties finding where the configurations are stored. 

To find where the configurations are stored, I tried to set a recognisable directory watch folder name, like `recognisable_watch_folder_name` and then searching through the contents of the files in`/home/username/` to see if I find that setting, such that I can modify it from a bash script.

However, since I was [not able to run a stable release of 7.11 on Ubuntu](https://github.com/Tribler/tribler/issues/6895), I downloaded [this](https://jenkins-ci.tribler.org/job/Build-Tribler_release/job/Build-Ubuntu64/522/artifact/tribler/build/debian/tribler_7.12.0-alpha.4-3-g0fb222e62_all.deb) `.deb file to run Tribler. That version 7.12.0-alpha.4-3 crashes if I go into settings and click a directory (e.g. the watch folder) on my instance of Ubuntu 22.04.

Hence, I would like to ask (either):

0. How can I set the watch folder to be: `/home/username/watchfolder` from the command line or a bash script? 
1. Is the watch folder usage boolean, and watch folder directory setting stored in some `config.yaml` file, and if yes, where?
2. Is there perhaps an CLI API that allows for specifying the Tribler watch folder?

**Comment by a-t-0** [comment]
# Approach I
I first inspected [the command line arguments for Tribler](https://github.com/Tribler/tribler/blob/main/src/run_tribler.py#L26). However, I did not find a setting for the watch folder. Next I searched the repository for "watch folder".

## Searching Watch Folder Setting
After inspecting the Tribler source code *I think* I  found the boolean that determines whether or not Tribler monitors a watch folder at [/src/tribler/core/components/watch_folder/settings.py](https://github.com/Tribler/tribler/blob/4287bb822fa6bb17cd555c12d38b54b594d2af27/src/tribler/core/components/watch_folder/settings.py#L7)
Along with the string for the watch directory in line 8 of that file:
```py
from tribler.core.config.tribler_config_section import TriblerConfigSection

# pylint: disable=no-self-argument


class WatchFolderSettings(TriblerConfigSection):
    enabled: bool = False
    directory: str = ''
```

## Changing Watch Folder Setting
I am not yet sure how to find that file when Tribler is installed using:
```
wget https://jenkins-ci.tribler.org/job/Build-Tribler_release/job/Build-Ubuntu64/522/artifact/tribler/build/debian/tribler_7.12.0-alpha.4-3-g0fb222e62_all.deb
    sudo dpkg -i tribler_7.12.0-alpha.4-3-g0fb222e62_all.deb
```
However, I will look into it.

# Approach II (successful)
The error I described when clicking on the folder icon in `Settings>General>Watch Folder` emerged only upon clicking the icon. So instead I copy pasted the textbox and clicked `save` in the top right of the settings. That did work. Next, I was able to search the directory `/home/username/` for any file with content: `recognisable_watch_folder_name`. That returned the file: `/home/username/.Tribler/7.12/triblerd.conf` which contained:
```
[watch_folder]
    enabled = True
    directory = /home/username/recognisable_watch_folder_name
[api]
    http_enabled = True
    http_port = some port
    key = some key
```

## Solution
So I think I would be able to pre-pend this: 
```
[watch_folder]
    enabled = True
    directory = /home/username/recognisable_watch_folder_name
[api]
```
to the `/home/username/.Tribler/7.12/triblerd.conf` file. I did not yet write a script that finds that file automatically regardless of the version and prepends that data.

**Comment by drew2a** [comment, marked as answer]
Hi!

> How can I set the watch folder to be: /home/username/watchfolder from the command line or a bash script?

You have to add the following lines to the `triblerd.conf` file:
```ini
[watch_folder]
    enabled = True
    directory =  /home/username/watchfolder 
```

> Is the watch folder usage boolean, and watch folder directory setting stored in some config.yaml file, and if yes, where?

We use an ini-like config file that is stored in `~/.Tribler/<TRIBLER_VERSION>/triblerd.conf` where `TRIBLER_VERSION` is Tribler version, like `7.11` or `7.12`.

> Is there perhaps an CLI API that allows for specifying the Tribler watch folder?

There is no special CLI API but you can use [Tribler's REST API](https://tribler.readthedocs.io/en/latest/restapi/introduction.html#overview) 

Still, the easiest way to change the config is to modify directly `triblerd.conf`.

The full specification of the config is here: https://github.com/Tribler/tribler/blob/d9c44f29487e48970a3b99adfe032afc3eef6c10/src/tribler/core/config/tribler_config.py#L34-L56

**Reply by a-t-0** [reply]
Thank you for your detailed solution, it was a cross-post where you beat me to the party! I also appreciate your reference to the Tribler REST API.
